src/eos-updater-avahi.c

```diff
@@ -0,0 +1,678 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ * Author: Krzesimir Nowak <krzesimir@kinvolk.io>
+ */
+
+#include "config.h"
+
+#include "eos-updater-avahi.h"
+
+#include "eos-util.h"
+
+#include "eos-updater-types.h"
+
+#include <avahi-client/client.h>
+#include <avahi-client/lookup.h>
+#include <avahi-common/address.h>
+#include <avahi-common/defs.h>
+#include <avahi-common/error.h>
+#include <avahi-common/malloc.h>
+#include <avahi-common/strlst.h>
+#include <avahi-glib/glib-malloc.h>
+#include <avahi-glib/glib-watch.h>
+
+#ifndef EOS_AVAHI_PORT
+#error "EOS_AVAHI_PORT is not defined"
+#endif
+
+static const gchar *const EOS_UPDATER_AVAHI_SERVICE_TYPE = "_eos_updater._tcp";
+static const gchar *const EOS_AVAHI_SERVICE_FILE_TEMPLATE =
+  "<service-group>\n"
+  "  <name replace-wildcards=\"yes\">EOS update service on %h</name>\n"
+  "  <service>\n"
+  "    <type>@TYPE@</type>\n"
+  "    <port>@PORT@</port>\n"
+  "    <txt-record>eos_txt_version=@TXT_VERSION@</txt-record>\n"
+  "    @MORE_TXT_RECORDS@\n"
+  "  </service>\n"
+  "</service-group>\n";
+
+static void
+eos_avahi_service_finalize_impl (EosAvahiService *service)
+{
+  g_free (service->name);
+  g_free (service->domain);
+  g_free (service->address);
+  g_strfreev (service->txt);
+}
+
+EOS_DEFINE_REFCOUNTED (EOS_AVAHI_SERVICE,
+                       EosAvahiService,
+                       eos_avahi_service,
+                       NULL,
+                       eos_avahi_service_finalize_impl)
+
+typedef enum
+{
+  EOS_AVAHI_DISCOVERING_AND_RESOLVING,
+  EOS_AVAHI_RESOLVING_ONLY,
+  EOS_AVAHI_FINISHED
+} EosAvahiState;
+
+struct _EosAvahiDiscoverer
+{
+  GObject parent_instance;
+
+  AvahiGLibPoll *poll;
+  AvahiClient *client;
+  AvahiServiceBrowser *browser;
+
+  EosAvahiDiscovererCallback callback;
+  gpointer user_data;
+  GDestroyNotify notify;
+  GMainContext *context;
+
+  GHashTable *discovered_services;
+  GPtrArray *found_services;
+  GError *error;
+  EosAvahiState state;
+  guint callback_id;
+};
+
+static void
+eos_avahi_discoverer_dispose_impl (EosAvahiDiscoverer *discoverer)
+{
+  guint callback_id = discoverer->callback_id;
+  gpointer user_data = discoverer->user_data;
+  GDestroyNotify notify = discoverer->notify;
+
+  discoverer->callback_id = 0;
+  if (callback_id > 0)
+    {
+      GSource* source;
+
+      source = g_main_context_find_source_by_id (discoverer->context,
+                                                 callback_id);
+      if (source != NULL)
+        g_source_destroy (source);
+    }
+
+  discoverer->user_data = NULL;
+  discoverer->notify = NULL;
+  if (user_data != NULL && notify != NULL)
+    notify (user_data);
```

Comment:
I'm not sure about it. I thought it might be convenient to avoid calling notify with the `NULL` user_data, so we can simply pass `g_object_unref` (which is not `NULL` safe) as `notify` function regardless of whether we pass `NULL` or some object as `user_data`.

If you wanted to call notify which does not care about the `user_data` at all, then you could just set the `user_data` to something like `GINT_TO_POINTER(1)`.

I'd need to look at the prior art in other libraries (like gio).